Mathlib/Computability/MyhillNerode.lean

```diff
@@ -0,0 +1,105 @@
+
+/-!
+# Myhill–Nerode theorem
+
+This file proves the Myhill–Nerode theorem via left quotients.
+
+Given a language `L` and a word `x`, the *left quotient* is the set of suffixes `y` such that
+`x ++ y` is in `L`. The *Myhill–Nerode theorem* shows that each left quotient, in fact, corresponds
+to the state of an automaton that matches `L`, and that `L` is regular if and only if there are
+finitely many such states.
+
+## Implementation notes
+
+Since mathlib doesn't have an `IsRegular` definition yet, we define regularity directly as the
+existence of a DFA.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Syntactic_monoid#Myhill%E2%80%93Nerode_theorem>
+-/
+
+universe u v
+variable {α : Type u} {σ : Type v} (L : Language α)
+
+namespace Language
+
+/-- The *left quotient* of `x` is the set of suffixes `y` such that `x ++ y` is in `L`. -/
+def leftQuotient (x : List α) : Language α := { y | x ++ y ∈ L }
+
+@[simp]
+theorem leftQuotient_nil : L.leftQuotient [] = L := rfl
+
+@[simp]
+theorem leftQuotient_append (x y : List α) :
+    L.leftQuotient (x ++ y) = (L.leftQuotient x).leftQuotient y := by
+  dsimp [leftQuotient, Language]
+  simp_rw [List.append_assoc]
+
+@[simp]
+theorem leftQuotient_mem (x y : List α) : y ∈ L.leftQuotient x ↔ x ++ y ∈ L := Iff.rfl
+
+theorem leftQuotient_accepts (M : DFA α σ) (x : List α) :
+    leftQuotient M.accepts x = M.acceptsFrom (M.eval x) := by
+  ext y
+  rw [DFA.mem_acceptsFrom, DFA.eval, ← DFA.evalFrom_of_append, ← DFA.mem_accepts, leftQuotient_mem]
+
+theorem leftQuotient_accepts' (M : DFA α σ) : leftQuotient M.accepts = M.acceptsFrom ∘ M.eval :=
+  funext <| leftQuotient_accepts M
+
+theorem finite_leftQuotient_of_dfa [Finite σ] (M : DFA α σ) :
+    Set.Finite (Set.range M.accepts.leftQuotient) :=
+  leftQuotient_accepts' M ▸
+    Set.finite_of_finite_preimage (Set.toFinite _)
+      (Set.range_comp_subset_range M.eval M.acceptsFrom)
+
+/-- The left quotients of a language are the states of an automaton that accepts the language. -/
+def toDFA : DFA α (Set.range L.leftQuotient) where
+  step s a := by
+    refine ⟨s.val.leftQuotient [a], ?_⟩
+    obtain ⟨y, hy⟩ := s.prop
+    exists y ++ [a]
+    rw [← hy, leftQuotient_append]
+  start := ⟨L, by exists []⟩
+  accept := { s | [] ∈ s.val }
+
+@[simp]
+theorem mem_toDFA_accept (s : Set.range L.leftQuotient) : s ∈ L.toDFA.accept ↔ [] ∈ s.val := Iff.rfl
+
+@[simp]
+theorem toDFA_step (s : Set.range L.leftQuotient) (a : α) :
+    (L.toDFA.step s a).val = s.val.leftQuotient [a] := rfl
+
+@[simp]
+theorem toDFA_start : L.toDFA.start.val = L := rfl
+
+@[simp]
+theorem toDFA_accepts : L.toDFA.accepts = L := by
+  ext x
+  rw [DFA.mem_accepts, ← DFA.eval]
+  suffices L.toDFA.eval x = L.leftQuotient x by simp [this]
+  induction x using List.list_reverse_induction with
+  | base => simp
+  | ind x a ih => simp [ih]
+
+theorem exists_dfa_of_finite_leftQuotient (h : Set.Finite (Set.range L.leftQuotient)) :
+    ∃ n, ∃ M : DFA α (Fin n), M.accepts = L :=
```

Comment:
I wish `∃ n, ∃ M : DFA α (Fin n), M.accepts = L` was a named proposition of `L` as its single argument. Dtto above.

Reply:
Yeah, I was afraid to say that directly, because `Language.IsRegular` should probably be defined via a general finite type, not necessarily `Fin n`.

Reply:
> because `Language.IsRegular` should probably be defined via a general finite type, not necessarily `Fin n`.

(I assume that's so we can prove `IsRegular` from a DFA directly, without having to go through `DFA α σ ≃ DFA α (Fin n)` first?)

Reply:
I mean it's not hard to show that wanting a fintype is the same as wanting specifically `Fin n` for some `n`, right? But it's hard enough that it should be in an API, and it could be the API for `IsRegular`.

Reply:
Thinking about this, I prefer the definition using `Fin n`, to avoid introducing an extra universe parameter for the states. (Of course we can have lemmas that convert to either.)

But we can discuss that in the PR that introduces the definition.

Reply:
I prefer to avoid extra universe parameters as well. There is a middle ground — define it for `{T : Type} [Fintype T]`.

Reply:
We introduced `IsRegular` in #13554, so I think the goal now is to rewrite this to reference the new definition.